In [1]:
import pandas as pd
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn import svm
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import GridSearchCV

In [2]:
train = pd.read_csv('/content/drive/MyDrive/datasets/SalaryData_Train(1).csv')
train

age          workclass  ...          native  Salary
0       39          State-gov  ...   United-States   <=50K
1       50   Self-emp-not-inc  ...   United-States   <=50K
2       38            Private  ...   United-States   <=50K
3       53            Private  ...   United-States   <=50K
4       28            Private  ...            Cuba   <=50K
...    ...                ...  ...             ...     ...
30156   27            Private  ...   United-States   <=50K
30157   40            Private  ...   United-States    >50K
30158   58            Private  ...   United-States   <=50K
30159   22            Private  ...   United-States   <=50K
30160   52       Self-emp-inc  ...   United-States    >50K

[30161 rows x 14 columns]

In [3]:
test = pd.read_csv('/content/drive/MyDrive/datasets/SalaryData_Test(1).csv')
test

age      workclass      education  ...  hoursperweek          native  Salary
0       25        Private           11th  ...            40   United-States   <=50K
1       38        Private        HS-grad  ...            50   United-States   <=50K
2       28      Local-gov     Assoc-acdm  ...            40   United-States    >50K
3       44        Private   Some-college  ...            40   United-States    >50K
4       34        Private           10th  ...            30   United-States   <=50K
...    ...            ...            ...  ...           ...             ...     ...
15055   33        Private      Bachelors  ...            40   United-States   <=50K
15056   39        Private      Bachelors  ...            36   United-States   <=50K
15057   38        Private      Bachelors  ...            50   United-States   <=50K
15058   44        Private      Bachelors  ...            40   United-States   <=50K
15059   35   Self-emp-inc      Bachelors  ...            60   United-States    >50K

[15060 rows x 14 columns]

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [5]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [6]:
train.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [7]:
col_cat = ['workclass', 'education', 'educationno', 'maritalstatus',
           'occupation', 'relationship', 'race', 'sex', 'native', 'Salary']

In [8]:
for i in col_cat:
    print(i, train[i].value_counts())

workclass  Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64
education  HS-grad         9840
 Some-college    6677
 Bachelors       5044
 Masters         1627
 Assoc-voc       1307
 11th            1048
 Assoc-acdm      1008
 10th             820
 7th-8th          557
 Prof-school      542
 9th              455
 12th             377
 Doctorate        375
 5th-6th          288
 1st-4th          151
 Preschool         45
Name: education, dtype: int64
educationno 9     9840
10    6677
13    5044
14    1627
11    1307
7     1048
12    1008
6      820
4      557
15     542
5      455
8      377
16     375
3      288
2      151
1       45
Name: educationno, dtype: int64
maritalstatus  Married-civ-spouse       14065
 Never-married             9725
 Divorced                  4214
 Separated                  939
 Widowed         

In [9]:
train1 = train.copy()
test1 = test.copy()

In [10]:
train1 = train1.drop('educationno', axis = 1)
test1 = test1.drop('educationno', axis = 1)

In [11]:
train1.columns

Index(['age', 'workclass', 'education', 'maritalstatus', 'occupation',
       'relationship', 'race', 'sex', 'capitalgain', 'capitalloss',
       'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [12]:
test1.columns

Index(['age', 'workclass', 'education', 'maritalstatus', 'occupation',
       'relationship', 'race', 'sex', 'capitalgain', 'capitalloss',
       'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [13]:
train1 = train1.apply(LabelEncoder().fit_transform)
train1.head()

age  workclass  education  ...  hoursperweek  native  Salary
0   22          5          9  ...            39      37       0
1   33          4          9  ...            12      37       0
2   21          2         11  ...            39      37       0
3   36          2          1  ...            39      37       0
4   11          2          9  ...            39       4       0

[5 rows x 13 columns]

In [14]:
test1 = test1.apply(LabelEncoder().fit_transform)
test1.head()

age  workclass  education  ...  hoursperweek  native  Salary
0    8          2          1  ...            39      37       0
1   21          2         11  ...            49      37       0
2   11          1          7  ...            39      37       1
3   27          2         15  ...            39      37       1
4   17          2          0  ...            29      37       0

[5 rows x 13 columns]

In [15]:
train1.shape

(30161, 13)

In [16]:
x_train = train1.iloc[:,:-1]
y_train = train1.iloc[:,-1]
print(x_train.head())
y_train.head()

   age  workclass  education  ...  capitalloss  hoursperweek  native
0   22          5          9  ...            0            39      37
1   33          4          9  ...            0            12      37
2   21          2         11  ...            0            39      37
3   36          2          1  ...            0            39      37
4   11          2          9  ...            0            39       4

[5 rows x 12 columns]


0    0
1    0
2    0
3    0
4    0
Name: Salary, dtype: int64

In [17]:
x_test = test1.iloc[:,:-1]
y_test = test1.iloc[:,-1]
print(x_test.head())
y_test.head()

   age  workclass  education  ...  capitalloss  hoursperweek  native
0    8          2          1  ...            0            39      37
1   21          2         11  ...            0            49      37
2   11          1          7  ...            0            39      37
3   27          2         15  ...            0            39      37
4   17          2          0  ...            0            29      37

[5 rows x 12 columns]


0    0
1    0
2    1
3    1
4    0
Name: Salary, dtype: int64

### Support Vector Machine Model Building

In [18]:
model = SVC()
model.fit(x_train , y_train)

SVC()

In [19]:
preds = model.predict(x_test)
acc = accuracy_score(y_test, preds) * 100

In [20]:
print('accuracy is', acc)
confusion_matrix(y_test, preds)

accuracy is 80.64409030544488


array([[11227,   133],
       [ 2782,   918]])

In [21]:
print('model accuracy is ', ((11227+918)/(11227+133+2782+918))*100)

model accuracy is  80.64409030544488


In [22]:
model1 = SVC(kernel = 'linear', gamma = 5, C = 0.1)
model1.fit(x_train , y_train)
preds1 = model1.predict(x_test)
acc1 = accuracy_score(y_test, preds1) * 100
print('accuracy is', acc1)
confusion_matrix(y_test, preds1)

accuracy is 79.19654714475432


array([[10866,   494],
       [ 2639,  1061]])

In [23]:
model2 = SVC(kernel = 'rbf', gamma = 10, C = 0.8)
model2.fit(x_train , y_train)
preds2 = model2.predict(x_test)
acc2 = accuracy_score(y_test, preds2) * 100
print('accuracy is', acc2)
confusion_matrix(y_test, preds2)

accuracy is 76.06241699867198


array([[11142,   218],
       [ 3387,   313]])

First I have load both train and test datasets. Then i have check for null values and data types. In this both datasets there is no null values. In both train and test data education and educationno columns are same so ihave drop the educationno in both train and test. After that by using labelencoder method i have converted all categorical to numeric. Then i have divided train data to xtrain and ytrain. And test data to xtest and ytest. Then i have build the model by using SVC by keeping kernel as default. Then by giving xtrain and ytrain i have fitted the model. after i have done model prediction by giving xtest and i have find the model accuracy with ytest and predictions. and model accuracy is 80.6%. Now i have build another model by giving kernel as linear , gamma = 5 and c = 0.1. And i have done predictions and model accuracy is 79%. And i have build another model by giving kernel as rbf, gamma = 10 and c = 0.8. Accuracy is 79%